src/poetry/utils/env/python_manager.py

```diff
@@ -0,0 +1,163 @@
+from poetry.utils.env.script_strings import GET_PYTHON_VERSION_ONELINER
+
+
+if TYPE_CHECKING:
+    from cleo.io.io import IO
+
+    from poetry.config.config import Config
+    from poetry.poetry import Poetry
+
+
+class Python:
+    def __init__(self, executable: str | Path, version: Version | None = None) -> None:
+        self.executable = Path(executable)
+        self._version = version
+
+    @property
+    def version(self) -> Version:
+        if not self._version:
+            if self.executable == Path(sys.executable):
+                _python_version = ".".join(str(v) for v in sys.version_info[:3])
+            else:
+                encoding = "locale" if sys.version_info >= (3, 10) else None
+                _python_version = decode(
+                    subprocess.check_output(
+                        [str(self.executable), "-c", GET_PYTHON_VERSION_ONELINER],
+                        text=True,
+                        encoding=encoding,
+                    ).strip()
+                )
+            self._version = Version.parse(_python_version)
+
+        return self._version
+
+    @cached_property
+    def patch_version(self) -> Version:
+        return Version.from_parts(
+            major=self.version.major,
+            minor=self.version.minor,
+            patch=self.version.patch,
+        )
+
+    @cached_property
+    def minor_version(self) -> Version:
+        return Version.from_parts(major=self.version.major, minor=self.version.minor)
+
+    @staticmethod
+    def _full_python_path(python: str) -> Path | None:
+        # eg first find pythonXY.bat on windows.
+        path_python = shutil.which(python)
+        if path_python is None:
+            return None
+
+        try:
+            encoding = "locale" if sys.version_info >= (3, 10) else None
+            executable = subprocess.check_output(
+                [path_python, "-c", "import sys; print(sys.executable)"],
+                text=True,
+                encoding=encoding,
+            ).strip()
+            return Path(executable)
+
+        except subprocess.CalledProcessError:
+            return None
+
+    @staticmethod
+    def _detect_active_python(io: IO) -> Path | None:
+        io.write_error_line(
+            "Trying to detect current active python executable as specified in"
+            " the config.",
+            verbosity=Verbosity.VERBOSE,
+        )
+
+        executable = Python._full_python_path("python")
+
+        if executable is not None:
+            io.write_error_line(f"Found: {executable}", verbosity=Verbosity.VERBOSE)
+        else:
+            io.write_error_line(
+                "Unable to detect the current active python executable. Falling"
+                " back to default.",
+                verbosity=Verbosity.VERBOSE,
+            )
+
+        return executable
+
+    @classmethod
+    def get_system_python(cls) -> Python:
+        return cls(executable=sys.executable)
```

Comment:
Is this the right name for this function? If Poetry is installed in a venv, e.g., using pipx, it'll pick up the pipx venv's Python binary, which seems surprising as a "system" Python version, which to me would imply "shared instance, outside any venv".

I don't see this function being used excepts in the tests, so I'm not clear what the intended behaviour is, but perhaps `get_running_python` or `get_poetry_python` would be more-accurate? (`get_preferred_python` open-codes the same behaviour so it will also pick up a pipx-poetry Python executable and not a system-shared instance, but it only appears to be used to extract the version and create new venvs, so that seems like the expected/intended behaviour there.)

Reply:
Good catch :+1: I guess my initial thought was, that there is a need to get the Python executable that poetry itself uses. But obviously the only place where this is required is within `get_preferred_python`, where we decide whether we use this executable or the one used in the current shell. We can either keep this method and used in within `get_preffered_python` to make the choice options more explicit or remove it.

Reply:
I kept this method and used in `get_preffered_python`.

Regarding the name: @Secrus & @radoering Are you fine with  `get_system_python` or do you prefer any other name?

Reply:
I understand the confusion.

On the one hand, `SystemEnv` is the env Poetry is installed in. `poetry env use system` tells Poetry to use it's own Python version as default for new venvs. Thus, in our wording `system` is Poetry's own env.

One the other hand, `virtualenvs.options.system-site-packages` does not refer to Poetry's own env. And most people will probably not associate `system` with Poetry's own env.

Not sure, if `get_poetry_python` is a better name or if it can be confused with the project's python. At least, `system` is quite consistent in our code. No strong opinion on this. 🤷

Reply:
For me `system python` was always confusing, but I agree with @radoering that it is quite consistent in the code now, so we can leave it as is.

Reply:
Good to see that we are on the same boat :smile: So I would leave it at it is for now.